File: constraints/scripts/planckCosmologyParametersConfig.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Cwd;
use XML::Simple;
use PDL;
use PDL::NiceSlice;
use PDL::LinearAlgebra;
use PDL::IO::Misc;
use PDL::MatrixOps;
use Data::Dumper;
use File::Temp qw(tempdir);

# Generates a configuration for cosmological parameters based on the Planck covariance matrix and suitable for use with the
# Galacticus MCMC constraints architecture. Parameters are expressed as linear combiantions of six independent normal deviates
# (labelled "cosmology0" through "cosmology5"). The output of this script can be included into the base parameter file for a
# Galacticus MCMC run, and actual cosmological parameters set by referencing those created in this file.
# Andrew Benson (16-October-2014)
 
# Get a temporary directory.
my $dir = tempdir( CLEANUP => 1 );

# Download Planck MCMC chains data set.
system("wget \"http://pla.esac.esa.int/pla-sl/data-action?COSMOLOGY.COSMOLOGY_OID=1200\" -O ".$dir."/COM_CosmoParams_base-plikHM-TT-lowTEB_R2.00.tar.gz")
    unless ( -e $dir."/COM_CosmoParams_base-plikHM-TT-lowTEB_R2.00.tar.gz" );

# Unpack Planck MCMC chains data set.
system("cd ".$dir."; tar xvfz COM_CosmoParams_base-plikHM-TT-lowTEB_R2.00.tar.gz")
    unless ( -e $dir."/base/plikHM_TT_lowTEB_lensing" );

# Specify Planck directory.
my $planckDirectoryName = $dir."/base/plikHM_TT_lowTEB_lensing/";

# Specify parameter names of interest.
my %parameterMap = 
    (
     "omegabh2"  => "planckCosmologyOmegaBaryon",
     "omegamh2*" => "planckCosmologyOmegaMatter",
     "tau"       => "planckCosmologyReionizationSuppressionOpticalDepth",
     "ns"        => "planckCosmologyPowerSpectrumIndex",
     "H0*"       => "planckCosmologyHubbleConstant",
     "sigma8*"   => "planckCosmologySigma8"
    );

# Read Planck parameter names.
my $parameters;
open(my $parameterNameFile,$planckDirectoryName."/base_plikHM_TT_lowTEB_lensing_post_BAO_H070p6_JLA.paramnames");
my $i = -1;
while ( my $line = <$parameterNameFile> ) {
    ++$i;
    $line =~ s/^\s*//;
    my @columns = split(/\s+/,$line);
    if ( exists($parameterMap{$columns[0]}) ) {
	$parameters->{$columns[0]}->{'column'} = $i+2;
	$parameters->{$columns[0]}->{'chain' } = pdl [];
    }
}
close($parameterNameFile);

# Check all required parameters were found.
foreach ( keys(%parameterMap) ) {
    die("planckCosmologyParametersConfig.pl: parameter '".$_."' not found")
	unless ( exists($parameters->{$_}) );
}

# Read Planck chains.
my @columnsToRead = map {$parameters->{$_}->{'column'}} sort(keys(%parameterMap));
opendir(my $planckDirectory,$planckDirectoryName);
while ( my $fileName = readdir($planckDirectory) ) {
    if ( $fileName =~ m/base_plikHM_TT_lowTEB_lensing_post_BAO_H070p6_JLA_\d+\.txt/ ) {
	my @columnData = rcols($planckDirectoryName."/".$fileName,@columnsToRead);
	# Append to chains.
	my $i = -1;
	foreach ( sort(keys(%parameterMap)) ) {
	    ++$i;
	    $parameters->{$_}->{'chain'} = $parameters->{$_}->{'chain'}->append($columnData[$i]);
	}
    }
}
closedir($planckDirectory);

# Find the means of the parameters.
my $parameterCount = scalar(keys(%parameterMap));
my $mean           = pdl zeroes($parameterCount);
$i                 = -1;
foreach ( sort(keys(%parameterMap)) ) {
    ++$i;
    $mean->(($i)) .= average($parameters->{$_}->{'chain'});
}

# Find the covariances of the parameters.
my $covariance = pdl zeroes($parameterCount,$parameterCount);
$i             = -1;
foreach my $iKey ( sort(keys(%parameterMap)) ) {
    ++$i;
    my$j = -1;
    foreach my $jKey ( sort(keys(%parameterMap)) ) {
	++$j;
	$covariance->(($i),($j)) .= 
	    average(
		($parameters->{$iKey}->{'chain'}-$mean->(($i)))
		*
		($parameters->{$jKey}->{'chain'}-$mean->(($j)))
	    );
    }
}

# Perform a Cholesky decomposition on the covariance matrix.
my $choleskyDecomposed = mchol($covariance);

# Construct the config data.
my $parameterConfig;
my $index = -1;
foreach my $parameter ( sort(keys(%parameterMap)) ) {
    # Find the index of this parameter in the vectors.
    ++$index;
    # Begin by setting the parameter equal to its mean value.
    my $value = "=".$mean->(($index))->string();
    # Loop over all independent normal deviates.
     for(my $i=0;$i<$parameterCount;++$i) {
	 # Extract the coefficient for this random deviate and (if it is non-zero) add it multiplied by the appropriate random
	 # deviate.
	 my $coefficient = $choleskyDecomposed->(($index),($i));
	 $value .= "+[cosmology".$i."]*".$coefficient
	     if ( $coefficient != 0.0 );
     }
    # Extract the parameter name, mapping to the Galacticus name if necessary.
    my $parameterName = $parameter;
    $parameterName = $parameterMap{$parameter}
          unless ( $parameterMap{$parameter} eq "" );
    # If the parameter is a density parameter then it has been multiplied by h^2 in the Planck analysis. Undo this.
    $value = "(".$value.")/(%[planckCosmologyHubbleConstant]/100.0)**2"
 	if ( $parameter =~ m/omega/ );
    # Add limits.
    $value = "min(".$value.",1.0)"
	if ( $parameter eq "omegamh2*" );
    $value = "max(".$value.",0.0)"
	if ( $parameter eq "tau" );    
    # Store the parameter config in our array.
    $parameterConfig->{$parameterName} = {"value" => $value};
}

# Output the configuration.
my $xmlOut = new XML::Simple (RootName=>"parameters");
print $xmlOut->XMLout($parameterConfig);

exit;

File: constraints/scripts/planckCosmologyParametersConfig.py
#!/usr/bin/env python3
# Generates a configuration for cosmological parameters based on the Planck covariance matrix and suitable for use with the
# Galacticus MCMC constraints architecture. Parameters are expressed as linear combinations of six independent normal deviates
# (labelled "cosmology0" through "cosmology5"). The output of this script can be included into the base parameter file for a
# Galacticus MCMC run, and actual cosmological parameters set by referencing those created in this file.
# Andrew Benson (16-October-2014)

import os
import re
import sys
import tempfile
import subprocess
import numpy as np
import xml.etree.ElementTree as ET
from xml.dom import minidom

# Specify parameter names of interest and their Galacticus equivalents.
parameter_map = {
    "omegabh2":  "planckCosmologyOmegaBaryon",
    "omegamh2*": "planckCosmologyOmegaMatter",
    "tau":       "planckCosmologyReionizationSuppressionOpticalDepth",
    "ns":        "planckCosmologyPowerSpectrumIndex",
    "H0*":       "planckCosmologyHubbleConstant",
    "sigma8*":   "planckCosmologySigma8",
}

with tempfile.TemporaryDirectory() as tmpdir:
    # Download Planck MCMC chains data set.
    tar_path = os.path.join(tmpdir, "COM_CosmoParams_base-plikHM-TT-lowTEB_R2.00.tar.gz")
    if not os.path.exists(tar_path):
        subprocess.run([
            "wget",
            "http://pla.esac.esa.int/pla-sl/data-action?COSMOLOGY.COSMOLOGY_OID=1200",
            "-O", tar_path,
        ], check=True)

    # Unpack Planck MCMC chains data set.
    planck_dir = os.path.join(tmpdir, "base", "plikHM_TT_lowTEB_lensing")
    if not os.path.exists(planck_dir):
        subprocess.run(["tar", "xvfz", tar_path], cwd=tmpdir, check=True)

    planck_dir = planck_dir + "/"

    # Read Planck parameter names to find the column index of each parameter in the chain files.
    paramnames_file = os.path.join(
        planck_dir,
        "base_plikHM_TT_lowTEB_lensing_post_BAO_H070p6_JLA.paramnames",
    )
    parameters = {}
    with open(paramnames_file) as f:
        for i, line in enumerate(f):
            cols = line.strip().split()
            if cols and cols[0] in parameter_map:
                # Chain files have weight and log-likelihood in columns 0-1; parameters start at column 2.
                parameters[cols[0]] = {"column": i + 2, "chain": []}

    # Check all required parameters were found.
    for key in parameter_map:
        if key not in parameters:
            sys.exit(f"planckCosmologyParametersConfig.py: parameter '{key}' not found")

    # Read Planck chain files, accumulating samples for each parameter.
    sorted_keys = sorted(parameter_map.keys())
    usecols = [parameters[k]["column"] for k in sorted_keys]
    chain_pattern = re.compile(
        r"base_plikHM_TT_lowTEB_lensing_post_BAO_H070p6_JLA_\d+\.txt$"
    )
    for fname in sorted(os.listdir(planck_dir)):
        if chain_pattern.match(fname):
            data = np.loadtxt(os.path.join(planck_dir, fname), usecols=usecols)
            if data.ndim == 1:
                data = data.reshape(1, -1)
            for i, key in enumerate(sorted_keys):
                parameters[key]["chain"].append(data[:, i])

    for key in sorted_keys:
        parameters[key]["chain"] = np.concatenate(parameters[key]["chain"])

    # Compute means and covariance matrix.
    param_count = len(parameter_map)
    mean = np.array([np.mean(parameters[k]["chain"]) for k in sorted_keys])
    covariance = np.zeros((param_count, param_count))
    for i, ki in enumerate(sorted_keys):
        for j, kj in enumerate(sorted_keys):
            covariance[i, j] = np.mean(
                (parameters[ki]["chain"] - mean[i])
                * (parameters[kj]["chain"] - mean[j])
            )

    # Perform Cholesky decomposition: covariance = L @ L.T
    L = np.linalg.cholesky(covariance)

    # Build XML output.
    root = ET.Element("parameters")
    for index, parameter in enumerate(sorted_keys):
        value = f"{mean[index]}"
        for i in range(param_count):
            coeff = L[index, i]
            if coeff != 0.0:
                value += f"+[cosmology{i}]*{coeff}"
        param_name = parameter_map[parameter]
        # Density parameters are stored as ω=Ωh² in Planck chains; convert to Omega.
        if "omega" in parameter:
            value = f"({value})/([planckCosmologyHubbleConstant]/100.0)**2"
        # Apply physical bounds.
        if parameter == "omegamh2*":
            value = f"min({value},1.0)"
        if parameter == "tau":
            value = f"max({value},0.0)"
        # Add an initial "=" so that Galacticus evaluates this parameter.
        value = "=" + value
        ET.SubElement(root, param_name, attrib={"value": value})

print(minidom.parseString(ET.tostring(root)).toprettyxml(indent="  "))
